Add pagination to changelog/blog main page

Skye967 opened the pull request:
## Type of change
When we are on the openstatus.dev/changelog or openstatus.dev/blog we added pagination to show only the last 10 entries.

<!--- What types of changes does your code introduce? Put an `x` in the box that apply: -->

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation

## Description

<!--- Describe your changes in detail -->

- I added a file for "use client" so it does not interfere with the metadata export
- created pagination function
- made next page and previous page handler function
- added buttons to flip pages and page number to the top and bottom of page

### A picture tells a thousand words (if any)

### Before this PR

<img width="1023" alt="before" src="https://github.com/user-attachments/assets/e00d7ac5-c28e-4c12-b624-98c1f8a44a4d">

### After this PR

<img width="1023" alt="after" src="https://github.com/user-attachments/assets/0a6e641b-125a-4dbd-914a-72947dab0880">


### Related Issue (optional)

<!--- Please link to the issue here: -->

[https://github.com/openstatusHQ/openstatus/issues/923](url)

Skye967 committed:
added pagination functionality to changelog/blog

Skye967 committed:
import cleanup

mxkaske reviewed (commented):
Appreciate your help on that one! 🔥 

A few thoughts on your implementation. It provides everything we need for client-side pagination. I think we can improve it by moving the state to a `searchParam` to allow server-side pagination. 

From a Ui perspective I'd just stick to our design system which is nothing less than shadcn ui and use the [pagination](https://ui.shadcn.com/docs/components/pagination) component at the end of the page. We can include the component within the `@openstatus/ui` package.

We can do everything in the `page.tsx` components to avoid creating new component.

What do you think of that?

Skye967 commented:
Thanks for the feedback, I think I did it that way because I also was have problems using next/navigation, I'll double check that and try put it on the server side. 
I'll change it to the shadcn ui component.

mxkaske commented:
> Thanks for the feedback, I think I did it that way because I also was have problems using next/navigation, I'll double check that and try put it on the server side. I'll change it to the shadcn ui component.

Sounds good! Feel free to point out your issue with `next/navigation` (either within the PR or via Discord DM).

Skye967 committed:
changed pagination fuctionality and ui

Skye967 committed:
Update page.tsx

some cleanup

Skye967 commented:
I think I did it how you suggested. I was going to use the shadcn ui, but I noticed there was some pagination components in the local directory. I decided on making the ListPagination component to utilize the pagination components which also had the right ui style consistency.

<img width="1077" alt="Screenshot 2024-07-13 at 10 28 06 PM" src="https://github.com/user-attachments/assets/b3871bfc-55b1-4872-8dd2-47824d81d42a">

mxkaske commented on apps/web/src/app/(content)/changelog/page.tsx:
```diff
@@ -40,13 +50,14 @@ export default async function Changelog() {
           <Button variant="outline" size="icon" asChild>
             <a href="/changelog/feed.xml" target="_blank" rel="noreferrer">
-              <Rss className="h-4 w-4" />
+              <span className="h-4 w-4">RSS</span>
```
Please keep the Icon.

mxkaske commented on apps/web/src/app/(content)/changelog/page.tsx:
```diff
@@ -40,13 +50,14 @@ export default async function Changelog() {
         }
       >
-        {changelogs.map((changelog) => (
+        <ListPagination current={currentPage} total={totalPages} />
```
Let's only use the pagination on the bottom of the page.

mxkaske commented on apps/web/src/components/ui/button.tsx:
We have the button component in `@openstatus/ui`. No need for the component here.

Skye967 replied:
I didn't create that component. I do not know why it said I did. They seem to be nearly Identical to each other. It'll probably be fine If it was deleted, should I delete it?

mxkaske commented on apps/web/src/components/ui/pagination.tsx:
Please add the component to `@openstatus/ui` (in packages/ui/src/..) instead of here.

Skye967 replied:
I also did not create this file either. I can move this one as well, It should be fine.

mxkaske commented on apps/web/src/app/(content)/changelog/page.tsx:
```diff
@@ -26,10 +27,19 @@ export const metadata: Metadata = {
-export default async function Changelog() {
-  const changelogs = allChangelogs.sort(
+export default function ChangelogClient(context: Context) {
+  const currentPage = parseInt(context.searchParams.page) || 1
```
Please use zod validation here. An example:

https://github.com/openstatusHQ/openstatus/blob/main/apps/web/src/app/app/%5BworkspaceSlug%5D/(dashboard)/monitors/(overview)/page.tsx#L20

(same for Blog page)

mxkaske commented on apps/web/src/app/(content)/changelog/page.tsx:
```diff
@@ -26,10 +27,19 @@ export const metadata: Metadata = {
 
-export default async function Changelog() {
-  const changelogs = allChangelogs.sort(
+export default function ChangelogClient(context: Context) {
```
```suggestion
export default function Changelog({
  searchParams,
}: {
  searchParams: { [key: string]: string | string[] | undefined };
}) {
```

mxkaske commented on apps/web/src/app/(content)/changelog/page.tsx:
```diff
@@ -1,17 +1,18 @@
 import { Timeline } from "@/components/content/timeline";
-import { Shell } from "@/components/dashboard/shell";
+import { allChangelogs } from "contentlayer/generated";
+import { Context } from "vm";
```
No need for that import

Skye967 replied:
Do is it not need the Shell component? It was there before

mxkaske replied:
I was referring to `import { Context } from "vm";` 😀

Skye967 replied:
Oh yeah. lol Yeah that's gone

mxkaske commented on apps/web/src/app/(content)/_components/ListPagination.tsx:
Please rename to `list-pagination` to match our file naming. Also avoid using `default export`

mxkaske commented on apps/web/package.json:
```diff
@@ -37,6 +37,7 @@
     "@openstatus/upstash": "workspace:*",
     "@openstatus/utils": "workspace:*",
     "@radix-ui/react-portal": "1.0.4",
+    "@radix-ui/react-slot": "1.0.2",
```
no need

mxkaske reviewed (commented):
Thanks @Skye967! Let me know if you have questions

Skye967 committed:
list-pagination adjustments and moved button and pagination to @openstatus/ui

Skye967 committed:
Merge branch 'add-pagination-to-changelog/blog-main-page' of https://github.com/Skye967/openstatus into add-pagination-to-changelog/blog-main-page

Skye967 committed:
list pagination adjustments

Skye967 committed:
Merge branch 'main' into add-pagination-to-changelog/blog-main-page

Skye967 commented:
<img width="1077" alt="Screenshot 2024-07-14 at 2 49 35 PM" src="https://github.com/user-attachments/assets/e694fd29-098c-4cd8-9e0b-bacb19ae9e95">

Adjusted the list-pagination component and added zod validation to the changelog and blog. I also made copies of the pagination and button components and put them in @openstatus/ui. I moved the list-pagination to the bottom of the pages.

mxkaske commented on apps/screenshot-worker/wrangler.toml:
```diff
@@ -14,6 +14,8 @@ browser = { binding = "MYBROWSER" }
 [[r2_buckets]]
 binding = "MY_BUCKET"
 bucket_name = "incident-screenshot"
+preview_bucket_name = "preview_bucket"
```
How comes adding that line?

Skye967 replied:
Sorry, I got rid of it. When I first ran the project I got some errors so I thought I needed it to run the project locally.

mxkaske commented on packages/ui/src/components/pagination.tsx:
Can we please use `Link` instead of `a`. See https://ui.shadcn.com/docs/components/pagination#nextjs

mxkaske reviewed (commented):
The build step is failing. Could you please have a look? 🙏

Skye967 committed:
Pagination Compnent changed a tag to next Link tag

Skye967 committed:
fix wrangler toml file

Skye967 committed:
some dependency cleanup

Skye967 committed:
changed web dir pagination a tag to Link tag

Skye967 commented:
I think the build step should be working. I can't test it properly because I have no Sentry auth token configured. I changed the pagination component to use the next Link tag. I had to install the dependency in the directory.

Skye967 committed:
Merge branch 'main' into add-pagination-to-changelog/blog-main-page

mxkaske commented:
Hey @Skye967, sorry for keeping you waiting. I'll be back from vacation end of the week and will have some time to follow up! I will test it locally and if that's ok for you I'll commit the last fixes on top of it! Thanks for your work so far.

mxkaske committed:
Merge branch 'main' into Skye976-add-pagination-to-changelog/blog-main-page

mxkaske committed:
chore: small stuff

mxkaske commented:
Hey @Skye967! Feel free to have a look at the final changes. Thanks for your work here and apologies it took so long. 🙏

mxkaske committed:
fix: lint

mxkaske committed:
fix: biome
